Replace PIO code with an inbuilt timer, based on https://learn.adafru…

mcarlson opened the pull request:
…it.com/intro-to-rp2040-pio-with-circuitpython/advanced-using-pio-to-drive-neopixels-in-the-background

Not really helping with the glitchiness I wonder if it's my hardware at this point?

Commit:
Replace PIO code with an inbuilt timer, based on https://learn.adafruit.com/intro-to-rp2040-pio-with-circuitpython/advanced-using-pio-to-drive-neopixels-in-the-background

blaz-r commented:
Many thanks for this. 

I'll test is as soon as I get the time. I'm not sure if this would cause any backwards-compatibility problems if merged to main, but if it works okay, I believe it probably shouldn't.

blaz-r commented on neopixel.py:
```diff
@@ -229,14 +249,25 @@ def __setitem__(self, idx, rgb_w):
         npix[15:21] = (255,0,0)     # <- sets 16,17 .. 20 to red
         npix[21:29:2] = (0,0,255)   # <- sets 21,23,25,27 to blue
         npix[1::2] = (0,0,0)        # <- sets all odd pixels to 'off'
+        npix[:] = [(0,5,0),(0,5,0)] # <- replaces all pixels with those from the array
         (the 'slice' cases pass idx as a 'slice' object, and
         set_pixel processes the slice)
 
         :param idx: Index can either be indexing number or slice
         :param rgb_w: Tuple of form (r, g, b) or (r, g, b, w) representing color to be used
         :return:
         """
-        self.set_pixel(idx, rgb_w)
+        if type(rgb_w) is list:
+            for i in range(self.num_leds):
+                self.set_pixel(i, rgb_w[i])
+        else:
+            self.set_pixel(idx, rgb_w)
+
+    def __len__(self):
+        return self.num_leds
+
+    def __getitem__(self, idx):
+        return self.get_pixel(idx)
```
Just a small comment here. Since this is part of other PR, could you maybe exclude it here?

blaz-r commented on neopixel.py:
```diff
@@ -85,23 +95,33 @@ def __init__(self, num_leds, state_machine, pin, mode="RGB", delay=0.0001):
         :param delay: [default: 0.0001] delay used for latching of leds when sending data
         """
         self.pixels = array.array("I", [0] * num_leds)
+        # self.pixels_out = array.array("I", [0] * num_leds)
         self.mode = mode
         self.W_in_mode = 'W' in mode
+        bpp = 3
         if self.W_in_mode:
             # RGBW uses different PIO state machine configuration
-            self.sm = rp2.StateMachine(state_machine, sk6812, freq=8000000, sideset_base=Pin(pin))
+            self.sm = rp2.StateMachine(state_machine, sk6812, freq=12_800_000, sideset_base=Pin(pin))
             # tuple of values required to shift bit into position (check class desc.)
             self.shift = ((mode.index('R') ^ 3) * 8, (mode.index('G') ^ 3) * 8,
                           (mode.index('B') ^ 3) * 8, (mode.index('W') ^ 3) * 8)
+            bpp = 4
         else:
-            self.sm = rp2.StateMachine(state_machine, ws2812, freq=8000000, sideset_base=Pin(pin))
+            self.sm = rp2.StateMachine(state_machine, ws2812, freq=12_800_000, sideset_base=Pin(pin))
```
Shouldn't this also have the new PIO code as only the sk6812 program is changed? 

If I understand this correctly, the program is now same for both strips, so maybe we can have just one program (but the pull_threshold will need to be adjusted for each state machine like in https://learn.adafruit.com/intro-to-rp2040-pio-with-circuitpython/advanced-using-pio-to-drive-neopixels-in-the-background).

A comment with link to original version this is based on should probably be included as well, to credit the authors of the original work.

blaz-r commented on neopixel.py:
```diff
@@ -85,23 +95,33 @@ def __init__(self, num_leds, state_machine, pin, mode="RGB", delay=0.0001):
         self.delay = delay
         self.brightnessvalue = 255
 
+        byte_count = bpp * num_leds
+        bit_count = byte_count * 8
+        padding_count = -byte_count % 4
+        self.header = bytearray(struct.pack("L", bit_count - 1))
+        self.trailer = bytearray(b"\0" * padding_count + struct.pack("L", 3840))
```
A comment here would be nice so we know why exactly this is needed in further development.

blaz-r reviewed (changes requested):
Thanks again for this 😄. Just some minor comments.

Commit:
Back out changes from https://github.com/blaz-r/pi_pico_neopixel/pull/19

Commit:
Add more comments and credits to the source

mcarlson commented:
I believe I've addressed your comments, let me know! Keep in mind the testcase I posted requires pull 19 to work. Also, I should mention the first four pixels are white for me when I apply this change, so it's not 100% yet.

blaz-r commented:
Thanks for addressing the comments. I'm not sure about the white pixel problem, but it might be due to the header and trailer because the code on that link I believe does some shifting and padding to fit the DMA requirements and I'm not sure if that's same here.

Commit:
Use a single array as a buffer, including header and trailer

Uses an offset to ensure get/set() happens in the right place. Also caches the memoryview once to make show() as efficient as possible. It turned out passing the header/trailer values as-is from python is the right approach. Everything works well for me now, maybe I'll look into DMA next :)

mcarlson commented:
You were right, that packing was bogus. I've updated the PR with the new code, it works really nicely now... Thanks for your patience!

Commit:
Remove references to delay as it's no longer needed

blaz-r commented:
Great to hear. Did the glitches also go away with this?

mcarlson commented:
Not entirely, though removing all interrupts and replacing with asyncio (especially Timers) seems to help a lot. It's waaay better though!

blaz-r commented:
Good. I'll probably have some time to review this and test next weekend. Thanks for all the effort 😄

blaz-r committed:
Merge branch 'main' into pio_timer

blaz-r commented:
Hi, I've finally had the time to check this, really sorry for the delay. It works quite well, but I had to change some code as rotation no longer works with this code.

The other problem is that this only works for RGBW, so merging this would mean that RGB strips are no longer supported. I'm not entirely sure how to fix this, as that would require rewriting PIO program to do the pull once 24 bits are read. I'll see what I can do.

mcarlson commented:
Ah right I was worried about that. I’ll see if I can pass another argument in and have it only push the first 3 bytes if set. I don’t have a normal run strip to test with though, I’ll order or see if I can borrow one!

-Max

On Mon, Apr 1, 2024 at 3:18 AM, Blaž Rolih ***@***.***(mailto:On Mon, Apr 1, 2024 at 3:18 AM, Blaž Rolih <<a href=)> wrote:

> Hi, I've finally had the time to check this, really sorry for the delay. It works quite well, but I had to change some code as rotation no longer works with this code.
>
> The other problem is that this only works for RGBW, so merging this would mean that RGB strips are no longer supported. I'm not entirely sure how to fix this, as that would require rewriting PIO program to do the pull once 24 bits are read. I'll see what I can do.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/blaz-r/pi_pico_neopixel/pull/20#issuecomment-2028881749), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AAABMCKAUN5ZTBC2MMPTYULY3BOSDAVCNFSM6AAAAABCZZDGGSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMRYHA4DCNZUHE).
> You are receiving this because you authored the thread.Message ID: ***@***.***>

blaz-r commented:
Thank you. I had an idea that could maybe work: when doing pull(ifempty) if we could somehow append some additional code to this ifempty and shift for 8 bits, then it should work. But at the moment I'm not sure if "ifempty" works outside pull instruction or/if it's possible to detect empty OSR.

blaz-r commented:
I think I figured it out. There is in fact and `osre` keyword that signals empty OSR 🥳

blaz-r committed:
Optimize and add WS2821 support

blaz-r committed:
Merge branch 'main' into pio_timer

blaz-r committed:
Remove unused imports

blaz-r committed:
Merge branch 'pio_timer' of https://github.com/mcarlson/pi_pico_neopixel into pio_timer

blaz-r commented:
When you get some time, test this code I pushed on your end (RGBW only, no need to order RGB). I tested on both RGB and RGBW and it works for me on all examples now.

blaz-r commented:
Looking at the new asm for ws2821, I think that the new logic introduces some additional low output, that might be problematic in some cases, although I had no issues with my strip - I'll try to fix it, but I won't be able to test for some time now.
